include/rx/ranges.hpp

```diff
@@ -965,12 +1011,18 @@ struct take {
         }
 
         [[nodiscard]] constexpr bool at_end() const noexcept {
-            return i == n || inner.at_end();
+            return i >= n || inner.at_end();
         }
 
         constexpr size_t size_hint() const noexcept {
             return n;
         }
+
+        constexpr void advance_by(size_t m) noexcept {
+            using RX_NAMESPACE::advance_by;
+            i += m;
+            advance_by(inner, m);
```

Comment:
```c++
if (i + m >= n || i + m > i) {
    using RX_NAMESPACE::advance_by;
    i += m;
    advance_by(inner, m);
} else {
    // The end is reached, or the addition would overflow, which implies the same.
    i = n;
}
```

Reply:
For the purposes of `tee` or other cases where `next()` has side effects, it is probably also necessary to call `advance_by` in the overflow case.